bin/e2hs-writer/src/reader.rs

```diff
@@ -0,0 +1,384 @@
+        stream! {
+        while current_block < self.ending_block {
+            if current_block < MERGE_BLOCK_NUMBER {
+                yield self.get_pre_merge_block_data(current_block);
+            } else if current_block < SHANGHAI_BLOCK_NUMBER {
+                yield self.get_pre_capella_block_data(current_block).await;
+            } else if current_block < CANCUN_BLOCK_NUMBER {
+                yield self.get_pre_deneb_block_data(current_block).await;
+            } else {
+                yield Err(anyhow!("Invalid block number: {current_block}"));
+            }
+            current_block += 1;
+        }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ethportal_api::types::{
+        consensus::{
+            beacon_block::{BeaconBlockBellatrix, BeaconBlockCapella},
+            beacon_state::BeaconState,
+            fork::ForkName,
+        },
+        execution::header_with_proof::{
+            build_historical_roots_proof, build_historical_summaries_proof,
+            BlockProofHistoricalRoots, BlockProofHistoricalSummaries,
+        },
+    };
+    use serde_yaml::Value;
+    use ssz::Decode;
+
+    use super::*;
+
+    #[rstest::rstest]
+    #[case(15539558, 4702208, 575, "block_proofs_bellatrix/beacon_block_proof-15539558-cdf9ed89b0c43cda17398dc4da9cfc505e5ccd19f7c39e3b43474180f1051e01.yaml")]
+    #[case(15547621, 4710400, 576, "block_proofs_bellatrix/beacon_block_proof-15547621-96a9313cd506e32893d46c82358569ad242bb32786bd5487833e0f77767aec2a.yaml")]
+    #[case(15555729, 4718592, 577, "block_proofs_bellatrix/beacon_block_proof-15555729-c6fd396d54f61c6d0f1dd3653f81267b0378e9a0d638a229b24586d8fd0bc499.yaml")]
+    #[tokio::test]
+    async fn test_pre_capella_proof_generation(
+        #[case] block_number: u64,
+        #[case] slot: u64,
+        #[case] _epoch: u64,
+        #[case] file_path: &str,
+    ) {
+        let test_vector = std::fs::read_to_string(format!(
+            "../../portal-spec-tests/tests/mainnet/history/headers_with_proof/{file_path}"
+        ))
+        .unwrap();
+        let test_vector: Value = serde_yaml::from_str(&test_vector).unwrap();
+        let actual_proof = BlockProofHistoricalRoots {
+            beacon_block_proof: serde_yaml::from_value(test_vector["beacon_block_proof"].clone())
+                .unwrap(),
+            beacon_block_root: serde_yaml::from_value(test_vector["beacon_block_root"].clone())
+                .unwrap(),
+            execution_block_proof: serde_yaml::from_value(
+                test_vector["execution_block_proof"].clone(),
+            )
+            .unwrap(),
+            slot: serde_yaml::from_value(test_vector["slot"].clone()).unwrap(),
+        };
+
+        let test_assets_dir =
+            format!("../../portal-spec-tests/tests/mainnet/history/headers_with_proof/beacon_data/{block_number}");
+        let hb_path = format!("{test_assets_dir}/historical_batch.ssz");
+        let hb_raw = std::fs::read(hb_path).unwrap();
```

Comment:
I am not sure what hb means

Reply:
It seems pretty obvious to me that it's `historical_batch` given the format string there...

Reply:
then this variable should be called `historical_batch`

Reply:
yeah, i made that change. but sometimes i do think there is a fair trade off between lengthy and simplified var names, given that the reader can understand the meaning without too much effort. especially in tests.